avm/src/vm/memory.rs

```diff
@@ -1,654 +1,390 @@
+  pub fn pop(self: &mut HandlerMemory, addr: i64) -> Result<(usize, i64), String> {
+    let mem = self.read_mut_fractal(addr);
+    if mem.len() > 0 {
+      return Ok(mem.pop().unwrap());
     } else {
-      let decision = arr.either_mem.pop().unwrap();
-      if decision < 0 {
-        let mut frac = HandlerMemory::new(None, 1);
-        frac.write_fixed(0, arr.mem.pop().unwrap());
-        // This is a really shitty side-channel signal that it's fixed data
-        frac.is_fixed = true;
-        return Ok(frac);
-      } else {
-        arr.mem.pop();
-        return Ok(arr.fractal_mem.pop().unwrap());
-      }
+      return Err("cannot pop empty array".to_string());
     }
   }
 
-  /// read address of string or fixed length data type and
-  /// return a reference to the data and its size
-  pub fn read_either(self: &HandlerMemory, addr: i64) -> (Vec<i64>, u8) {
-    if addr < 0 && self.is_neg_addr_gmem(addr) {
-      // assume it's fixed memory
-      return (vec![self.read_fixed(addr)], 8);
-    } else if addr < 0 { // It's closure memory
-      return self.closure_args[0].read_either(addr - CLOSURE_ARG_MEM_START);
-    }
-    // first check if the data is stored in a pointer
-    if self.registers_ish[addr as usize] > 0 {
-      return if self.registers_ish[addr as usize] == HM_PTR_ADDR {
-        let var = self.read_fractal_mem(addr);
-        (var, 0)
-      } else {
-        (vec![self.read_fixed(addr)], 8)
-      }
-    }
-    // test if the data read is itself a string/array
-    return if self.either_mem[addr as usize] > -1 {
-      let var = self.read_fractal_mem(addr);
-      (var, 0)
-    } else {
-      // Nope, it's fixed data. We can safely read 8 bytes for all of the fixed types
-      (vec![self.read_fixed(addr)], 8)
-    };
+  pub fn register(self: &mut HandlerMemory, addr: i64, orig_addr: i64) {
+    let (a, b) = self.addr_to_idxs(orig_addr);
+    self.set_addr(addr, a, b);
   }
 
-  pub fn read_fixed(self: &HandlerMemory, addr: i64) -> i64 {
-    if addr < 0 {
-      // global memory
-      if self.is_neg_addr_gmem(addr) {
-        let a = (0 - addr - 1) as usize;
-        let result = i64::from_ne_bytes((&self.gmem[a..a+8]).try_into().unwrap());
-        return result;
-      }
-      // closure arguments memory
-      let closure = &self.closure_args[0];
-      return closure.read_fixed(addr - CLOSURE_ARG_MEM_START);
-    }
-    let reg = self.registers_ish[addr as usize];
-    if reg == NORMAL_ADDR {
-      unsafe {
-        return self.mem.as_ptr().add(addr as usize).read();
-      }
-    } else if reg == FX_PTR_ADDR {
-      unsafe {
-        let ptr = usize::from_ne_bytes(self.mem[addr as usize].to_ne_bytes()) as *const i64;
-        let val = *ptr;
-        return val;
-      }
-    } else {
-      panic!("Trying to get fixed value from fractal pointer");
-    }
+  pub fn register_in(self: &mut HandlerMemory, orig_addr: i64, fractal_addr: i64, offset_addr: i64) {
```

Comment:
nit - register_in_fractal?

Reply:
I think that would make it more confusing? Would `register_in_fractal` only work if the value inside is also a fractal?

I can add comments, though.

`register` - Just aliases one address for another.
`register_in` - Points to a value stored inside of a nested memory address
`register_out` Creates a pointer in the nested memory address to the outer memory address (inverse of register_in)

Reply:
got it, those comments make things a lot clearer